File: InterviewPrep-Python/CrackingTheCodingInterview/Chapter1-ArraysAndStrings/Q8_NullifyMatrix.py
'''
Created on Dec 29, 2016

@author: drawkward
'''
import unittest

def nullifyRows(matrix, row):
    for i in range(len(matrix)):
        matrix[row][i]=0

def nullifyColumns(matrix, column):
    for i in range(len(matrix)):
        matrix[i][column]=0
        
def nullifyMatrix(matrix):
    
    if(len(matrix)==0 or len(matrix)!=len(matrix[0])):
        return False
    
    rows=[]
    columns = []
    
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            if(matrix[i][j]==0):
                rows.append(i)
                columns.append(j)
    
    for row in rows:
        nullifyRows(matrix, row)
        
    for column in columns:
        nullifyColumns(matrix, column)   
        
    return matrix         

class Test(unittest.TestCase):

    data = [
        ([
            [1, 2, 3, 4, 0],
            [6, 0, 8, 9, 10],
            [11, 12, 13, 14, 15],
            [16, 0, 18, 19, 20],
            [21, 22, 23, 24, 25]
        ], [
            [0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0],
            [11, 0, 13, 14, 0],
            [0, 0, 0, 0, 0],
            [21, 0, 23, 24, 0]
        ])
    ]

    def test_zero_matrix(self):
        for [test_matrix, expected] in self.data:
            actual = nullifyMatrix(test_matrix)
            self.assertEqual(actual, expected)



if __name__ == "__main__":
    unittest.main()

File: InterviewPrep-Java/src/CrackingTheCodingInterview/Chapter1_ArraysAndStrings/Q8_NullifyMatrix.java
package CrackingTheCodingInterview.Chapter1_ArraysAndStrings;

public class Q8_NullifyMatrix {

	public static void nullifyRow(int matrix[][], int row){
		for(int i=0;i<matrix.length;i++){
			matrix[row][i]=0;
		}
	}
	
	public static void nullifyColumn(int matrix[][], int column){
		for(int i=0;i<matrix.length;i++){
			matrix[i][column] = 0;
		}
	}
	
	public static boolean nullifyMatrix(int matrix[][]){
		
		if(matrix.length==0 || matrix.length!=matrix[0].length) return false;
		boolean row[] = new boolean[matrix.length];
		boolean column[] = new boolean[matrix.length];
		
		for(int i=0;i<matrix.length;++i){
			for(int j=0;j<matrix[0].length;++j){
				if(matrix[i][j]==0){
					row[i] = true;
					column[j] = true;
				}
			}
		}
		for(int i =0;i<row.length;++i){
			if(row[i])
				nullifyRow(matrix, i);
		}
		
		for(int i =0;i<column.length;++i){
			if(column[i])
				nullifyColumn(matrix, i);
		}
		return true;
	}
	
	public static void main(String args[]){
		int matrix[][] = {{1,2,3,4},
						{0,2,3,5},
						{1,5,6,7},
						{5,4,3,0}		
		};
		
		nullifyMatrix(matrix);
		for(int i=0;i<4;++i){
			for(int j=0;j<4;++j){
				System.out.print(matrix[i][j]);
			}
			System.out.println();
		}
	}
}
